extensions/amp-story-360/0.1/amp-story-360.js

```diff
@@ -568,14 +568,46 @@ export class AmpStory360 extends AMP.BaseElement {
    * @private
    */
   onDeviceOrientation_(e) {
+    const smoothFactor = 0.1;
+
+    if (!this.smoothA_) {
+      this.smoothA_ = e.alpha;
+    } else {
+      this.smoothA_ =
+        e.alpha * smoothFactor + this.smoothA_ * (1 - smoothFactor);
+    }
```

Comment:
The issue with this is when we cross a loop and sensor value jumps from ~180 to ~-180 and vice versa. The moving average algorithm will cause the POV to pan backwards brutally over a few frames to the new sensor value.

In order to avoid this we need to constantly prevent these ~360 degree gaps so that we move from 180 to 180 instead of -179. Do do that, we can monitor the sensor jumps and when they occurs, offset the sensor value by 360 degrees in the correct direction, like so:

```js
    if (this.smoothA_ === undefined) {
      this.smoothA_ = e.alpha;
    } else {
      if ((Math.sign(e.alpha) != Math.sign(this.prevSensorAlpha_)) && Math.abs(e.alpha) > 170) {
        // We have a -180 / +180 loop. Add or substract 360 degrees from the angle.
        this.alphaOffset_ += e.alpha < this.prevSensorAlpha_ ? 1 : -1;
      }
      this.prevSensorAlpha_ = e.alpha; // We need frame-level check for the sign change.
      this.smoothA_ =
        (e.alpha + 360 * this.alphaOffset_) * smoothFactor 
        + this.smoothA_ * (1 - smoothFactor);
    }
```

with `alphaOffset_` initialized at zero. `alphaOffset_` effectively becomes a loop counter (and you can break the logic by spinning on yourself in the same direction `Number.MAX_SAFE_INTEGER + 1` times 😅)

Reply:
This is great but I'm wondering if we can average the values of the `Matrix.rotation`
It works well when the rotation is calculated with raw sensor data, so we could average the previous and next `Matrix.rotation`.